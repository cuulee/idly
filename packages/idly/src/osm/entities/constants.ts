interface IGeometries {
  readonly POINT: 'point';
  readonly VERTEX: 'vertex';
  readonly AREA: 'area';
  readonly LINE: 'line';
  readonly RELATION: 'relation';
}

export const Geometries: IGeometries = {
  POINT: 'point',
  VERTEX: 'vertex',
  AREA: 'area',
  LINE: 'line',
  RELATION: 'relation'
};

export type Geometry = IGeometries[keyof IGeometries];
