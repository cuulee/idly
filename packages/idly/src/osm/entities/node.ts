import { List, Record } from 'immutable';

import { genLngLat, LngLat } from 'osm/geo_utils/lng_lat';
import { Properties, propertiesGen } from 'osm/others/properties';
import { ITags, tagsFactory } from 'osm/others/tags';

export type Id = string;
export type Version = number;
export type Visible = boolean;

export class Node extends Record({
  id: 'n-0',
  loc: genLngLat([NaN, NaN]),
  tags: tagsFactory(),
  type: 'node',
  properties: propertiesGen()
}) {
  public id: Id;
  public tags: ITags;
  public type: string;
  public loc: LngLat;
  public properties: Properties;
  public set(k: string, v: any): Node {
    return super.set(k, v) as Node;
  }
}

export function nodeFactory(obj: {
  id: Id;
  tags?: ITags;
  loc?: LngLat;
  properties?: Properties;
}) {
  return new Node(obj);
}
