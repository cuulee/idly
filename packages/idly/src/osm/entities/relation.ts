import { List, Map, Record } from 'immutable';
import { ITags, tagsFactory } from 'src/osm/others/tags';

import { Properties, propertiesGen } from 'src/osm/others/properties';

type Id = string;
type Version = number;
type Visible = boolean;

export class Relation extends Record({
  id: 'w-0',
  properties: propertiesGen(),
  tags: tagsFactory(),
  type: 'relation',
  members: List()
}) {
  public id: Id;
  public type: string;
  public tags: ITags;
  public properties: Properties;
  public members: List<Map<string, any>>;
  public set(k: string, v: any): Relation {
    return super.set(k, v) as Relation;
  }
}

export function relationFactory(obj: {
  id: Id;
  tags?: ITags;
  properties?: Properties;
  members?: List<Map<string, any>>;
}) {
  return new Relation(obj);
}
