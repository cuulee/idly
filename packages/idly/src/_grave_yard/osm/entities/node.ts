import { Record } from 'immutable';

import { Geometry } from 'osm/entities/constants';
import { Properties, propertiesGen } from 'osm/entities/helpers/properties';
import { Tags, tagsFactory } from 'osm/entities/helpers/tags';
import { genLngLat, LngLat } from 'osm/geo_utils/lng_lat';

export type Id = string;
export type Version = number;
export type Visible = boolean;

/**
 * @BUG cant use enum Geometries
 *  Need more investigation, for now
 *  converted it to an object.
 * @REVISIT I think this ^^ bug was due
 *  to cyclic dependencies which I have fixed now.
 *  Enum should in paper work.
 */
console.log('circular dep check', Geometry.POINT);
export class Node extends Record({
  id: 'n-0',
  loc: genLngLat([NaN, NaN]),
  tags: tagsFactory(),
  type: 'node',
  properties: propertiesGen(),
  /**
   * @TOFIX urgent remove this asap
   */
  geometry: Geometry.POINT
}) {
  readonly id: Id;
  readonly tags: Tags;
  readonly type: string;
  readonly loc: LngLat;
  readonly properties: Properties;
  public set(k: string, v: any): Node {
    return super.set(k, v) as Node;
  }
}

export function nodeFactory(
  obj: {
    id: Id;
    tags?: Tags;
    loc?: LngLat;
    properties?: Properties;
  },
  parentWays?
) {
  return new Node(obj);
}
