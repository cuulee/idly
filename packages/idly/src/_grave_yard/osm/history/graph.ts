import { Map, Record, Set } from 'immutable';

import { Entities, Entity, EntityId } from 'osm/entities/entities';
import { Node } from 'osm/entities/node';
import { Relation } from 'osm/entities/relation';
import { Way } from 'osm/entities/way';
import { groupBy } from 'ramda';

export class Graph extends Record({
  node: Map(),
  way: Map(),
  relation: Map()
}) {
  // Map entries are for finding a particular
  // entity which could be a heavy computation
  public node: Map<string, Node>;
  public way: Map<string, Way>;
  public relation: Map<string, Relation>;
  public set(k: string, v: Node | Way | Relation): Graph {
    return super.set(k, v) as Graph;
  }
  public getEntity(e: EntityId): Entity {
    if (e[0] === 'n') return super.getIn(['node', e]);
    if (e[0] === 'w') return super.getIn(['way', e]);
    if (e[0] === 'e') return super.getIn(['relation', e]);
    throw new Error(`unknown entity got ${e}`);
  }
}
type EntityType = Node | Way | Relation;

export function graphFactory(entitiesList: EntityType[] = []) {
  const createMapFromArray = (array = []) => {
    const map = Map();
    return map.withMutations(m => {
      array.forEach(a => {
        m.set(a.id, a);
      });
    });
  };
  const { node, way, relation } = groupBy(e => e.type, entitiesList);

  return new Graph({
    node: createMapFromArray(node),
    way: createMapFromArray(way),
    relation: createMapFromArray(relation)
  });
}

export function graphFactoryFromSet(entitiesSet: Entities = Set()) {
  const createDeepMap = (set: any = Set()) => {
    const map = Map();
    return map.withMutations(m => {
      set.forEach(a => {
        m.set(a.id, a);
      });
    });
  };
  const { node, way, relation } = entitiesSet.groupBy(e => e.type).toObject();
  // : { node: Node; way: Way; relation: Relation }
  return new Graph({
    node: createDeepMap(node),
    way: createDeepMap(way),
    relation: createDeepMap(relation)
  });
}
