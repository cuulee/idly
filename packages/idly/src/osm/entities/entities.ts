import { Set } from 'immutable';
import { Node } from 'osm/entities/node';
import { Relation } from 'osm/entities/relation';
import { Way } from 'osm/entities/way';

export type Entities = Set<Node | Way | Relation>;
export type Entity = Node | Way | Relation;
