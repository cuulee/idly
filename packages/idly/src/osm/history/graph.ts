import { Record, Map, List } from 'immutable';
import { ITags, tagsFactory } from 'src/osm/others/tags';
import { Node } from 'src/osm/entities/node';
import { Way } from 'src/osm/entities/way';
import { Relation } from 'src/osm/entities/relation';
import { groupBy } from 'ramda';
import { Properties, propertiesGen } from 'src/osm/others/properties';

var graphBaseRecord = Record({
  node: Map(),
  way: Map(),
  relation: Map()
});

export class Graph extends graphBaseRecord {
  node: Map<string, Node>;
  way: Map<string, Way>;
  relation: Map<string, Relation>;
  set(k: string, v: any): Graph {
    return <Graph>super.set(k, v);
  }
}
type EntityType = Node | Way | Relation;

export function graphFactory(entitiesList: EntityType[] = []) {
  const createMapFromArray = (array = []) => {
    var map = Map();
    return map.withMutations(m => {
      array.forEach(a => {
        m.set(a.id, a);
      });
    });
  };
  var { node, way, relation } = groupBy(e => e.type, entitiesList);
  return new Graph({
    node: createMapFromArray(node),
    way: createMapFromArray(way),
    relation: createMapFromArray(relation)
  });
}
