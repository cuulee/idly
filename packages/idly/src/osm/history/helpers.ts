import { Map } from 'immutable';

import { Graph } from 'osm/history/graph';
import { graphFactory, graphFactoryFromSet } from 'osm/history/graph';

import { Entities, EntitiesId } from 'osm/entities/entities';
import { Node } from 'osm/entities/node';
import { Relation } from 'osm/entities/relation';
import { Way } from 'osm/entities/way';

type EntitiesMap = Map<string, Node | Way | Relation>;
type Entity = Node | Way | Relation;

export function graphSetEntity(graph: Graph, entity: Entity): Graph {
  return graph.setIn([entity.type, entity.id], entity) as Graph;
}

export function graphSetEntitiesSet(graph: Graph, entities: Entities): Graph {
  const patchGraph = graphFactoryFromSet(entities);
  const mergedGraph = graph.mergeWith(
    (oldV, newV) => oldV.merge(newV),
    patchGraph
  ) as Graph;
  return mergedGraph;
}

export function graphSetEntities(graph: Graph, entities: Entity[]): Graph {
  const patchGraph = graphFactory(entities);
  const mergedGraph = graph.mergeWith(
    (oldV, newV) => oldV.merge(newV),
    patchGraph
  ) as Graph;
  return mergedGraph;
}

export function graphRemoveEntities(graph: Graph, entities: Entity[]): Graph {
  return graph.withMutations(g => {
    entities.forEach(e => {
      g.removeIn([e.type, e.id]);
    });
  }) as Graph;
}

export function graphRemoveEntitiesWithId(
  graph: Graph,
  entities: EntitiesId
): Graph {
  return graph.withMutations(g => {
    entities.forEach(e => {
      if (e[0] === 'n') g.removeIn(['node', e]);
      else if (e[0] === 'w') g.removeIn(['way', e]);
      else g.removeIn(['relation', e]);
    });
  }) as Graph;
}

export function graphRemoveEntity(graph: Graph, entity: Entity): Graph {
  return graph.removeIn([entity.type, entity.id]) as Graph;
}

export function calculateParentWays(graph: Graph): any {
  let parentWays = Map();
  parentWays = parentWays.withMutations(map => {
    graph.way.forEach(e => {
      e.nodes.forEach(n => {
        map.set(n, e.id);
      });
    });
  });
  return parentWays;
}
