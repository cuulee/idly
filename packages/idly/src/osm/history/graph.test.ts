import { List, Map, Set } from 'immutable';

import { Node, nodeFactory } from 'osm/entities/node';
import { relationFactory } from 'osm/entities/relation';
import { wayFactory } from 'osm/entities/way';
import { graphFactory } from 'osm/history/graph';
import {
  graphRemoveEntities,
  graphRemoveEntity,
  graphSetEntities,
  graphSetEntity
} from 'osm/history/helpers';

describe('Graph Factory', function() {
  // it('returns a new graph if self is frozen', function () {
  //     var graph = graphFactory();
  //     expect(graph.update()).not.toBe(graph);
  // });
  //   var graph = graphFactory({

  //   });
  const node = nodeFactory({ id: 'n1' });
  const way = wayFactory({ id: 'w-1', tags: Map({ foo: 'bar' }) });
  const relation = relationFactory({ id: 'r1' });
  it('returns self if empty', function() {
    const graph = graphFactory([node, way]);
    expect(graph).toMatchSnapshot();
    expect(graph.node).toEqual(Map({ n1: node }));
    expect(graph.way).toEqual(Map({ 'w-1': way }));
  });

  it('ignore multiple same copies', function() {
    const graph = graphFactory([node, way, node]);
    expect(graph.node.size).toEqual(1);
    expect(graph.way.size).toEqual(1);
  });
  it('ignore multiple same id', function() {
    const n2 = nodeFactory({ id: 'n1', tags: Map({ foo: 'foo' }) });
    const graph = graphFactory([node, way, n2]);
    expect(graph.node.size).toEqual(1);
    expect(graph.way.size).toEqual(1);
  });
  it('adds multiple nodes', function() {
    const n2 = nodeFactory({ id: 'n2', tags: Map({ foo: 'foo' }) });
    const graph = graphFactory([node, way, n2]);
    expect(graph.node.size).toEqual(2);
    expect(graph.way.size).toEqual(1);
  });
  it('adds multiple relations', function() {
    const r1 = relationFactory({ id: 'r1', tags: Map({ foo: 'foo' }) });
    const r2 = relationFactory({ id: 'r2', tags: Map({ foo: 'foo' }) });

    const graph = graphFactory([node, r2, way, r1]);
    expect(graph.node.size).toEqual(1);
    expect(graph.way.size).toEqual(1);
    expect(graph.relation.size).toEqual(2);
  });

  it('adds multiple relations', function() {
    const r1 = relationFactory({ id: 'r1', tags: Map({ foo: 'foo' }) });
    const r2 = relationFactory({ id: 'r2', tags: Map({ foo: 'foo' }) });

    const graph = graphFactory([relation, node, r2, way, r1]);
    expect(graph).toMatchSnapshot();
  });
});

// describe.skip('#remove', function() {
//   const node = nodeFactory({ id: 'n1' });
//   const graph = graphFactory({
//     entities: Map({ n1: nodeFactory({ id: 'n1' }) })
//   });

//   it('returns a new graph', function() {
//     var node2 = nodeFactory({ id: 'n2' });
//     var graph1 = graphSetEntity(graph, node2);
//     expect(graphRemoveEntity(graph1, node2)).toEqual(graph);
//   });

//   it("doesn't modify the receiver", function() {
//     expect(graph.entities.get(node.id)).toEqual(node);
//   });

//   it('removes non existing entity', function() {
//     var node2 = nodeFactory({ id: 'n2' });
//     expect(
//       graphRemoveEntity(graph, node2).entities.get(node2.id)
//     ).toBeUndefined();
//   });

//   it('removes the entity as a parentWay', function() {
//     // var node = new Node({ id: "n" }),
//     var node2 = nodeFactory({ id: 'n2' });
//     var w1 = wayFactory({ id: 'w', nodes: List(['n']) });
//     var graph2 = graphSetEntities(graphFactory({}), List([node2, w1]));

//     expect(graph.remove(w1).parentWays(node)).toEqual([]);
//   });

//   it('removes the entity as a parentRelation', function() {
//     var node = new Node({ id: 'n' }),
//       r1 = Relation({ id: 'w', members: [{ id: 'n' }] }),
//       graph = graphFactory([node, r1]);
//     expect(graph.remove(r1).parentRelations(node)).toEqual([]);
//   });
// });
