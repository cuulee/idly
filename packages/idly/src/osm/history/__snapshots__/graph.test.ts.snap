// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph Factory adds multiple relations 1`] = `
Object {
  "node": Object {
    "n1": Object {
      "geometry": "point",
      "id": "n1",
      "loc": Object {
        "lat": NaN,
        "lon": NaN,
      },
      "properties": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": 0,
        "visible": true,
      },
      "tags": Object {},
      "type": "node",
    },
  },
  "relation": Object {
    "r1": Object {
      "geometry": "relation",
      "id": "r1",
      "members": Array [],
      "properties": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": 0,
        "visible": true,
      },
      "tags": Object {
        "foo": "foo",
      },
      "type": "relation",
    },
    "r2": Object {
      "geometry": "relation",
      "id": "r2",
      "members": Array [],
      "properties": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": 0,
        "visible": true,
      },
      "tags": Object {
        "foo": "foo",
      },
      "type": "relation",
    },
  },
  "way": Object {
    "w-1": Object {
      "geometry": "line",
      "id": "w-1",
      "nodes": Array [],
      "properties": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": 0,
        "visible": true,
      },
      "tags": Object {
        "foo": "bar",
      },
      "type": "way",
    },
  },
}
`;

exports[`Graph Factory returns self if empty 1`] = `
Object {
  "node": Object {
    "n1": Object {
      "geometry": "point",
      "id": "n1",
      "loc": Object {
        "lat": NaN,
        "lon": NaN,
      },
      "properties": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": 0,
        "visible": true,
      },
      "tags": Object {},
      "type": "node",
    },
  },
  "relation": Object {},
  "way": Object {
    "w-1": Object {
      "geometry": "line",
      "id": "w-1",
      "nodes": Array [],
      "properties": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": 0,
        "visible": true,
      },
      "tags": Object {
        "foo": "bar",
      },
      "type": "way",
    },
  },
}
`;
